### PR TITLE
rewrites the admin token if it wasn't updated

### LINKED ISSUE #2127: master "edit settings" broken
**User Story**

Setting the adminToken to "" in cluster responses in cf2d7bb91bb7a54f804023eec45b4b7a952cb89c seems to have broken the "Edit Settings" (editProviderSettings()) feature in the UI, which now posts a cluster object with empty adminToken to /api/v3/dc/dbl1/cluster/{cluster}, resulting in a 400 response ("invalid cluster: invalid admin token: token is malformed, must match [bcdfghjklmnpqrstvwxz2456789]{6}\\.[bcdfghjklmnpqrstvwxz2456789]{16}").

We're using 2.7.7 currently.

**Acceptance criteria**

"Edit Settings" in the dashboard should work.

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: passing cluster resource without admin token to API
will no longer cause "invalid cluster: invalid admin token" error

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2127

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: api/pkg/handler/cluster.go [hunk 1]
```diff
@@ -164,6 +164,9 @@ func updateClusterEndpoint(cloudProviders map[string]provider.CloudProvider) end
 		newCluster.ObjectMeta = oldCluster.ObjectMeta
 		newCluster.Status = oldCluster.Status
 		newCluster.Spec.Cloud = kubermaticapiv1.UpdateCloudSpec(newCluster.Spec.Cloud, oldCluster.Spec.Cloud)
+		if len(newCluster.Address.AdminToken) == 0 {
```
Comment (p0lyn0mial) [inline api/pkg/handler/cluster.go:167]: fixes `invalid cluster: invalid admin token` error.

File: api/pkg/crd/kubermatic/v1/cluster.go [hunk 2]
```diff
@@ -199,24 +199,46 @@ func UpdateCloudSpec(updatedShared, existingShared CloudSpec) CloudSpec {
...
+			updated.Azure.ClientID = existingShared.Azure.ClientID
+		}
+		if len(updated.Azure.ClientSecret) == 0 {
+			updated.Azure.ClientSecret = existingShared.Azure.ClientSecret
+		}
 	}
-	if updated.Openstack != nil && len(updated.Openstack.Username) == 0 && len(updated.Openstack.Password) == 0 {
-		updated.Openstack.Username = existingShared.Openstack.Username
-		updated.Openstack.Password = existingShared.Openstack.Password
+	if updated.Openstack != nil {
```
Comment (p0lyn0mial) [inline api/pkg/crd/kubermatic/v1/cluster.go:218]: fixes I want to update only the password, not the password AND the user name

**Review by multi-io — COMMENTED** [review]
Ah, so you basically keep giving cluster responses with empty token to the client, but when the client posts an update, replace the empty token with the existing one for the cluster. I'll update our stage to this solution, test it and give feedback asap

Update: Fix works in our stage cluster.

**Review by multi-io — APPROVED** [review]
LGTM

**Review by mrIncompetent — APPROVED** [review]

**Comment by p0lyn0mial** [comment]
@multi-io many thanks for reviewing and testing.

**Pull request merged**